exercises/dominik_arominski/ex02_oo_basics/src/main.cpp

```diff
@@ -0,0 +1,154 @@
+#include <iostream>
+#include <string>
+#include <cmath>
+#include <vector>
+#include <algorithm>
+
+class Shape {
+
+protected:
+  std::string type;
+  int nSides;
+
+public:
+  
+  std::string GetType() const { return type;};
+  int GetSides() const { return nSides;};
+  
+  virtual double CalculatePerimeter() const { return 0;};  
+  virtual double CalculateArea() const { return 0;};  
+
+  friend std::ostream& operator<<(std::ostream& out, const Shape& shape) {
```

Comment:
Why friend?

Reply:
You are correct, it does not have to be friend of this class as it accesses only public fields and functions of this class. I removed it from the body of the class.